doc/en/monkeypatch.rst

```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
+will be raised if the set/deletion operation has no target.
+
+Consider the following scenarios:
+
+1. Modifying the behavior of a function or the property of a class for a test e.g.
+there is an API call or database connection you will not make for a test but you know
+what the expected output should be. Use :py:meth:`monkeypatch.setattr` to patch the
+function or property with your desired testing behavior. This can include your own functions.
+Use :py:meth:`monkeypatch.delattr` to remove the function or property for the test.
+
+2. Modifying the values of dictionaries e.g. you have a global configuration that
+you want to modify for certain test cases. Use :py:meth:`monkeypatch.setitem` to patch the
+dictionary for the test. :py:meth:`monkeypatch.delitem` can be used to remove items.
+
+3. Modifying environment variables for a test e.g. to test program behavior if an
+environment variable is missing, or to set multiple values to a known variable.
+:py:meth:`monkeypatch.setenv` and :py:meth:`monkeypatch.delenv` can be used for
+these patches.
+
+4. Use :py:meth:`monkeypatch.syspath_prepend` to add to the system ``$PATH`` and
+:py:meth:`monkeypatch.chdir` to change the context of the current working directory
+during a test.
+
 See the `monkeypatch blog post`_ for some introduction material
 and a discussion of its motivation.
 
 .. _`monkeypatch blog post`: http://tetamap.wordpress.com/2009/03/03/monkeypatching-in-unit-tests-done-right/
 
-
 Simple example: monkeypatching functions
 ----------------------------------------
 
-If you want to pretend that ``os.expanduser`` returns a certain
-directory, you can use the :py:meth:`monkeypatch.setattr` method to
-patch this function before calling into a function which uses it::
+Consider a scenario where you are working with user directories. In the context of
+testing, you do not want your test to depend on the running user. ``Monkeypatch``
+can be used to patch functions dependent on the user to always return a
+specific value.
+
+In this example, :py:meth:`monkeypatch.setattr` is used to patch ``os.path.expanduser``
+so that the known testing string ``"/abc"`` is always used when the test is run.
+This removes any dependency on the ``$USER`` environment variable for testing purposes.
+:py:meth:`monkeypatch.setattr` must be called before the function which will use
+the patched function is called.
+After the test function finishes the ``os.path.expanduser`` modification will be undone.
 
-    # content of test_module.py
+.. code-block:: python
+
+    # Content of test_module.py with source code and the test.
+    # os.path is imported for reference in monkeypatch.setattr().
     import os.path
-    def getssh(): # pseudo application code
-        return os.path.join(os.path.expanduser("~admin"), '.ssh')
 
-    def test_mytest(monkeypatch):
+
+    def getssh():
+        """Simple function to return expanded homedir ssh path from $USER"""
+        user = os.getenv("$USER")
```

Comment:
this looks up the environment variable named '$USER` -- I think you mean `os.getenv('USER')` (which could return `None`)

The `os.path.expanduser` call for "the current user" would just be:

```python
    return os.path.join(os.path.expanduser("~"), ".ssh")
```

no need for `getenv` or string formatting